gflownet/utils/batch.py

```diff
@@ -668,22 +686,71 @@ def get_logprobs(
             valid. The flag is False if the corresponding logprob value is a zero / None
             placefolder.
         """
+        if not self._logprobs_available:
+            self._compute_logprobs()
 
         if backward:
-            if self.logprobs_backward[0] is None:
-                return None, None
-            logprobs = tfloat(
-                self.logprobs_backward, float_type=self.float, device=self.device
-            )
-            valids = tbool(self.logprobs_backward_valid, device=self.device)
+            return self.logprobs_backward, self.logprobs_backward_valid
         else:
-            if self.logprobs_forward[0] is None:
-                return None, None
-            logprobs = tfloat(
-                self.logprobs_forward, float_type=self.float, device=self.device
+            return self.logprobs_forward, self.logprobs_forward_valid
+
+    def _compute_logprobs(self):
+        """
+        Convert all logprobs and their validity flags to tensors and update validity flags
+        """
+        self.logprobs_forward, self.logprobs_forward_valid = (
+            self._logprobs_to_clean_tensor(
+                self.logprobs_forward, self.logprobs_forward_valid
+            )
+        )
+        self.logprobs_backward, self.logprobs_backward_valid = (
+            self._logprobs_to_clean_tensor(
+                self.logprobs_backward, self.logprobs_backward_valid
             )
-            valids = tbool(self.logprobs_forward_valid, device=self.device)
-        return logprobs, valids
+        )
+        self._logprobs_available = True
+
+    def _logprobs_to_clean_tensor(self, logprobs: List, logprobs_valid: List[bool]):
+        """
+        Convert logprobs and logprobs_valid to tensors and update validity flags
+
+        Parameters
+        ----------
+        logprobs: list
+            A list of logprobs collected over a sequence of add_to_batch calls. The
+            elements could be None or a single-number torch.tensor attached or detached
+            from a computations graph collected to the loss
+        logprobs_valid: list of bools
+            A list of validity flags collected over a sequence of add_to_batch calls which
+            indicate whether the corresponding element in the logprobs list is a valid
+            logprob (True) or not (False). When the flag is False, the corresponding logprob
+            is either None or torch.tensor(0.0) detached from the computational graph.
+
+        Returns
+        -------
+        (logprobs_clean, logprobs_valid): a tuple of 1-d tensors
+            - logprobs_clean is a 1-d float tensor with logprobs, where all valid logprobs
+            remain connected to their computational graphs and all non-valid logprobs are zeros.
+            - logprobs_valid is a 1-d boolean tesor with flags indicating whether the corresponding
+            element in the logprobs is a valid logprob (True) or not (False).
```

Comment:
When multiple variables are returned (indeed, as a tuple), the way to list them in the docstrings is one by one. I will make this change myself.

Reply:
https://github.com/alexhernandezgarcia/gflownet/pull/396/commits/4a67c0d08362bbdc166a753d959e9fcc4c95f211